wfdb/io/datasource.py

```diff
@@ -0,0 +1,93 @@
+from enum import Enum, auto
+from pathlib import Path
+from typing import Dict
+from urllib.parse import urlparse
+
+
+class DataSourceType(Enum):
+    LOCAL = auto()
+    HTTP = auto()
+
+
+class DataSource:
+    def __init__(self, name: str, ds_type: DataSourceType, uri: str):
+        self.name = name
+        self.ds_type = ds_type
+        self.uri = uri
+
+    def __str__(self):
+        return f"{self.name} : {self.ds_type} : {self.uri}"
+
+    @property
+    def uri(self):
+        return self._uri
+
+    @uri.setter
+    def uri(self, value: str):
+        if self.ds_type == DataSourceType.LOCAL:
+            path = Path(value)
+            if not path.is_absolute():
+                raise ValueError(
+                    "uri field for a LOCAL DataSource must be a valid absolute path"
+                )
+        elif self.ds_type is DataSourceType.HTTP:
+            url = urlparse(value)
+            if not url.netloc:
+                raise ValueError(
+                    "uri field for an HTTP DataSource must be a valid URL"
+                )
+        self._uri = value
+
+
+_PHYSIONET_DATA_SOURCE = DataSource(
+    "physionet",
+    DataSourceType.HTTP,
+    "http://physionet.org/content/",
+)
+
+# Dict of configured data sources
+_data_sources: Dict[str, DataSource] = {"physionet": _PHYSIONET_DATA_SOURCE}
```

Comment:
Should _data_sources be all caps? Feels slightly odd to have this variable hanging around here.

Reply:
It's a global (module level) variable for storing the data sources. Not a constant.